### PR TITLE
Fix minor random bias

### PR DESCRIPTION
**Pull request by vcsjones**
This corrects two uses of RandomNumberGenerator that resulted in slight bias in their results. After some internal discussion it was determined that the bias did not weaken the randomness to the point where this needed to be treated as a vulnerability.

This also adds some tests just to make sure the random generation is not completely broken.

### CODE REVIEW

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs [hunk 1]
```diff
@@ -947,30 +947,61 @@ private static void ValidateTrustAttribute(TRUSTED_DOMAIN_INFORMATION_EX domainI
...
 
-            return password;
+                // Create a mask for only the applicable bits.
+                int mask = range;
+                mask |= mask >> 1;
+                mask |= mask >> 2;
+                mask |= mask >> 4;
+                // Don't need >> 8 or >> 16 since it is known the range is less than 255.
+
+                byte[] randomByte = new byte[1];
```
Comment (stephentoub) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs:991]: this is now allocating a new array per byte?

Comment (vcsjones) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs:991]: Hm, yeah. It's only for framework targets that don't have Span friendly random APIs, but on the other hand it's easy to just pull it out of the local method, so, fixed.

File: src/libraries/System.DirectoryServices/tests/System/DirectoryServices/ActiveDirectory/TrustHelperTests.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            HashSet<string> passwords = new HashSet<string>();
+
+            for (int i = 0; i < 100; i++)
+            {
+                string password = (string)createTrustPasswordMethod.Invoke(null, null);
+
+                // password is a 15 character length string from a 87 character set.
+                // log2(87)*15 is about 96, the odds of which we will generate a duplicate
+                // here are next to nothing.
+                Assert.True(passwords.Add(password), "password is a duplicate");
```
Comment (bartonjs) [inline src/libraries/System.DirectoryServices/tests/System/DirectoryServices/ActiveDirectory/TrustHelperTests.cs:33]: If this fails, it'll print
```
password is a duplicate
   Expected: True
       Actual: False
```


Assert.True's string needs to be what you're asserting, not what went wrong.

File: src/libraries/System.Net.Http/tests/UnitTests/DigestAuthenticationTests.cs [hunk 3]
```diff
@@ -104,5 +105,25 @@ public async Task DigestResponse_ShouldSendQop(string response, string match, st
...
+            HashSet<string> cnonces = new HashSet<string>();
+
+            for (int i = 0; i < 100; i++)
+            {
+                string token = (string)mi.Invoke(null, null);
+
+                // cnonce is a 16 character length string from a 62 character set.
+                // log2(62)*16 is about 95, the odds of which we will generate a duplicate
+                // here are next to nothing.
+                Assert.True(cnonces.Add(token), "cnonce is a duplicate");
```
Comment (bartonjs) [inline src/libraries/System.Net.Http/tests/UnitTests/DigestAuthenticationTests.cs:125]: Same "the message, sadly, is backwards"

**Review by bartonjs — APPROVED** [review]

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs [hunk 4]
```diff
@@ -947,30 +947,64 @@ private static void ValidateTrustAttribute(TRUSTED_DOMAIN_INFORMATION_EX domainI
...
+                do
+                {
+                    rng.GetBytes(randomBuffer);
+                    result = mask & randomBuffer[0];
+                }
+                while (result > range);
+
+                Debug.Assert(result < toExclusive);
+                return result;
+            }
```
Comment (stephentoub) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs:1006]: There's no way to achieve the desired goals here without implementing a custom non-biased GetInt32 method?

Comment (vcsjones) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs:1006]: @stephentoub For .NET Standard 2.0, not to my knowledge. Based on an off-GitHub conversation with @bartonjs I think he was on the same page that we'd have to lift and adapt `GetInt32`.

Comment (bartonjs) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs:1006]: Well, we could trim the alphabet, but that speeds up brute force (especially if we trim it down to 1 :smile:).

The alternative to dropping this copy here is something like

```C#
byte[] random = new byte[SomeNumber];
char[] output = new char[Whatever];
int randomPos = random.Length;

for (int writePos = 0; writePos < output.Length; writePos++)
{
    int randomVal;

    while (true)
    {
        if (randomPos >= random.Length)
        {
            RandomNumberGenerator.GetBytes(random);
            randomPos = 0;
        }

        randomVal = random[randomPos];
        randomPos++;
 
        Debug.Assert(alphabet.Length > 64 && alphabet.Length <= 128);
        randomVal &= 127;

        if (randomVal < alphabet.Length)
        {
            break;
        }
    }

    output = alphabet[randomVal];
}
```
Which amounts to the same thing.

File: src/libraries/System.DirectoryServices/tests/System/DirectoryServices/ActiveDirectory/TrustHelperTests.cs (hunk 2 above)
Comment (am11) [reply src/libraries/System.DirectoryServices/tests/System/DirectoryServices/ActiveDirectory/TrustHelperTests.cs]: Nice thing to know. :)
Not an expert, but I believe a custom assertion message should also try to capture complete context, something like: "expected distinct passwords but found a duplicate".

**Comment by vcsjones** [comment]
@joperezr Not sure what's up with the enterprise-linux leg but it seems unrelated. I think this is okay to merge.

**Comment by bartonjs** [comment]
I haven't hit merge because I don't know what's in the enterprise-linux leg and I've been hoping that its DNS failures would get worked out.  If you (@joperezr) are fine ignoring the leg, then I am, too.  (I'll admit, I have difficulty understanding what sort of test could be failing due to this change)

**Comment by stephentoub** [comment]
> I haven't hit merge because I don't know what's in the enterprise-linux leg and I've been hoping that its DNS failures would get worked out. 

I believe it's fine to ignore.  @wfurt can confirm.

**Comment by wfurt** [comment]
correct. The `enterprise-linux` failure is tracked by https://github.com/dotnet/core-eng/issues/15594 and it is environmental regression caused by infrastructure changes.

**Pull request merged**